aregion.cpp

```diff
@@ -2345,8 +2349,81 @@ ARegion *ARegionList::FindNearestStartingCity(ARegion *start, int *dir)
 	return 0;
 }
 
-int ARegionList::GetPlanarDistance(ARegion *one, ARegion *two,
-		int penalty, int maxdist)
+// Some structures for the get_connected_distance function
+// structures to allow us to do an efficient search
+struct RegionVisited {
+	int x, y, z;
+	bool operator==(const RegionVisited &v) const { return x == v.x && y == v.y && z == v.z; }
+ };
+class RegionVisitHash {
+public:
+	size_t operator()(const RegionVisited v) const {
+		return std::hash<uint32_t>()(v.x) ^ std::hash<uint32_t>()(v.y) ^ std::hash<uint32_t>()(v.z);
+	}
+};
+struct QEntry { ARegion *r; int dist; };
+class QEntryCompare {
+public:
+	// We want to sort by min distance
+	bool operator()(const QEntry &below, const QEntry &above) const { return above.dist < below.dist; }
+};
+
+// This doesn't really need to be on the ARegionList but, it's okay for now.
+int ARegionList::get_connected_distance(ARegion *start, ARegion *target, int penalty, int maxdist) {
```

Comment:
Have you considered using functions in the `graphs.h` or extending that file with the new features?

Reply:
I think that A* should go to the `graphs.h` in one way or another and use stuff there.

Reply:
I hadn't, but I will.  I looked at them this morning and they are a bit complex in some ways more than I think was needed, but I also kind of see why you did it that way.   I also need them to actually return the distance/exit early if the target isn't within the distance, so it might need another 'variant' of those.  Either way, yes, it's possible to revamp them to use it, but I might want to defer that to a followup PR/work after I get the other things I *must* have for the current game (victory condition work and map generation work).  Are you ok with deferring it?